src/org/github/jamm/MemoryMeter.java

```diff
@@ -1,250 +1,241 @@
+            assert current != null;
+            long size = measure(current);
+            listener.objectMeasured(current, size);
+            total += size;
+
+            if (current instanceof Object[]) {
+                addArrayChildren((Object[]) current, stack, tracker, listener);
+            } else if (current instanceof ByteBuffer && omitSharedBufferOverhead) {
+                total += ((ByteBuffer) current).remaining();
+            } else {
+                addFieldChildren(current, stack, tracker, listener);
+            }
+        }
+
+        listener.done(total);
+        return total;
     }
 
-    public Builder unbuild() {
-        return new Builder(guess,
-                           byteBufferMode,
-                           ignoreOuterClassReference,
-                           ignoreKnownSingletons,
-                           ignoreNonStrongReferences,
-                           ignoreDontMeasure
-        );
+    private void addFieldChildren(Object current, Deque<Object> stack, Set<Object> tracker, MemoryMeterListener listener) {
+        Class<?> cls = current.getClass();
+        while (cls != null) {
+            for (Field field : cls.getDeclaredFields()) {
+                if (fieldFilters.ignore(field)) {
+                    continue;
+                }
+
+                field.setAccessible(true);
+                Object child;
+                try {
+                    child = field.get(current);
+                } catch (IllegalAccessException e) {
+                    throw new RuntimeException(e);
+                }
+
+                if (child != null && !tracker.contains(child)) {
+                    stack.push(child);
+                    tracker.add(child);
+                    listener.fieldAdded(current, field.getName(), child);
+                }
+            }
+
+            cls = cls.getSuperclass();
+        }
     }
 
-    static final byte BB_MODE_NORMAL = 0;
-    static final byte BB_MODE_OMIT_SHARED = 1;
-    static final byte BB_MODE_SHALLOW = 2;
-    static final byte BB_MODE_HEAP_ONLY_NO_SLICE = 3;
+    private void addArrayChildren(Object[] current, Deque<Object> stack, Set<Object> tracker, MemoryMeterListener listener) {
+        for (int i = 0; i < current.length; i++) {
+            Object child = current[i];
+            if (child != null && !tracker.contains(child)) {
+
+                Class<?> childCls = child.getClass();
+                if (classFilter.ignore(childCls)) {
+                    continue;
+                }
+
+                stack.push(child);
+                tracker.add(child);
+                listener.fieldAdded(current, Integer.toString(i) , child);
+            }
+        }
+    }
 
-    public static class Builder {
-        private Guess guess = Guess.BEST;
-        private byte byteBufferMode = BB_MODE_NORMAL;
+    /**
+     * Builder for {@code MemoryMeter} instances
+     */
+    public static final class Builder {
```

Comment:
Can we add a bit more info around the different parameters, Javadoc or in the ReadMe as we have very specific cases  like the `IgnoreOuterClassReference`?
Probably the best in the readme in my opinion